### PR TITLE
micah/70/query string

### LINKED ISSUE #65: Include Thumbnail Filter State in URL
User Story:

>As a Blockbuilder Search user, I want the thumbnail filter state (the boolean corresponding to the checkbox) to be part of the URL, so that I can share links to results pages that only include thumbnails.

Desired behavior:

 * Clicking the checkbox for including thumbnails should update the URL
 * The checkbox should be initialized based on the flag presence in the URL

Related: #5 #29

### LINKED ISSUE #70: make url hash more human-readable
make url hash more human-readable

what if instead of this 

`http://[::]:8889/search#textt%3Dforce`

we had this:

`http://[::]:8889/search#text=force`

?

### PR DESCRIPTION
**Pull request by micahstubbs**
This PR encodes the search options in a web-standards query string.  It fixes the impedance mismatch that existed before, where we stored search options client-side in the location hash.

This PR also add debouncing for the main search text and username text input fields, so after typing is complete, the search is run automatically.  Likewise, the d3 API, d3 modules, and d3 version select boxes (dropdown menus) also trigger a new search automatically when their values change.

Also, following @erikhazzard's recommendation, this PR uses `history.pushState()` to update the page's url without triggering a page reload

docs on `history.pushState()` https://developer.mozilla.org/en-US/docs/Web/API/History_API#Example_of_pushState()_method

docs on `URLSearchParams` https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

This PR also persists thumbnail filter state in the url. Now our search bar inputs are closer to having consistent behavior.

This PR is a better alternative to the url hash hacking in https://github.com/enjalot/blockbuilder-search/pull/71

fix #70
fix #65 

## before 

`http://[::]:8889/search#text%3Dforce`

## after

`http://[::]:8889/search?text=force`

## live demo
here's an ngrok tunnel to a server running on a dev machine.  you should be able to try out the new behavior here:

https://1c3aa83f.ngrok.io/search

you should try some complex queries:

https://1c3aa83f.ngrok.io/search?text=bar&user=curran&api=d3.select%2Cd3.csv&d3modules=d3-legend&d3version=v3

### CODE REVIEW

**Comment by micahstubbs** [comment]
@curran I like this approach with query strings much better.  it feels more natural to use the url building and parsing methods that the browser API provides.  I also think we avoid some of the edge cases that @erikhazzard mentioned earlier about the other window.location.hash approach https://github.com/enjalot/blockbuilder-search/pull/72

File: src/components/app.js [hunk 1]
```diff
@@ -4,39 +4,58 @@ import { bindActionCreators } from 'redux'
...
     if (hash) {
       const options = hash.slice(1).split(';')
-      const object = {}
       options.forEach(option => {
-        const keyvalue = option.split('=')
-        object[keyvalue[0]] = keyvalue[1]
+        const pair = option.split('=')
+        const key = pair[0]
+        const value = pair[1]
+        updateQueryString(key, value)
```
Comment (micahstubbs) [inline src/components/app.js]: again, we maintain backwards compatability with old, location hash based search links.

you can try an old one against the new code on the dev server like this:

https://1c3aa83f.ngrok.io/search#user=enjalot;d3version=v4

Comment (micahstubbs) [reply src/components/app.js]: @curran notice how the hash is automatically upgraded to a query string when you visit that link ✨

File: src/components/searchbar.js [hunk 2]
```diff
@@ -239,18 +275,11 @@ const SearchBar = React.createClass({
...
-  componentDidUpdate() {
-    if (this.refs) {
-      if (this.refs.search || this.refs.search === '') {
-        this.refs.search.value = this.props.query.text
-      }
-      if (this.refs.user && this.props.query.userRaw) {
-        this.refs.user.value = this.props.query.userRaw
-      }
-    }
-  },
```
Comment (micahstubbs) [inline src/components/searchbar.js]: not removing this, just moving it up to the top of the component's class methods.

**Comment by erikhazzard** [comment]
That was fast, very nice!  🎉

**Comment by curran** [comment]
Woohoo! This is fantastic! Thanks @micahstubbs for your work here.

File: src/components/app.js (hunk 1 above)
Comment (curran) [reply src/components/app.js]: Fantastic!

File: src/components/app.js [hunk 3]
```diff
@@ -4,39 +4,75 @@ import { bindActionCreators } from 'redux'
...
+        case 'd3version':
+          query.d3version = value
+          break
+        case 'api':
+          query.api = value.split(',')
+          break
+        case 'd3modules':
+          query.d3modules = value.split(',')
+          break
+        case 'thumb':
```
Comment (curran) [inline src/components/app.js]: Very nice work. Looks great!

File: src/util/debounce.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export default function debounce(fn, ms) {
+  let timeout
+  // eslint-disable-next-line func-names
```
Comment (curran) [inline src/util/debounce.js]: I'm seeing a lot of `eslint-disable-next-line`, which makes me think perhaps the linting settings are too aggressive.

Comment (micahstubbs) [reply src/util/debounce.js]: yea, I was thinking the same.  I've created this issue to review our linting settings https://github.com/enjalot/blockbuilder-search/issues/73

**Comment by micahstubbs** [comment]
and deployed https://blockbuilder.org/search?thumb=true&user=curran 🎉 

screen shot 2018-10-11 at 10 47 20 am

**Pull request merged**